src/ruby/puppetserver-lib/puppet/server/logger.rb

```diff
@@ -27,5 +27,25 @@ def handle(msg)
 class Puppet::Server::Logger
   def self.init_logging
     Puppet::Util::Log.newdestination(:logback)
+    Puppet[:log_level] = level_from_logback(get_logger)
```

Comment:
If I'm not mistaken, I think the trouble with this approach is that you can only set the log level once, when we start up. Future changes to the logback loglevel won't be reflected by the puppet loglevel.

I think we want something that queries the logback level when a log message is generated, so you can change it any time and have it be reflected across server & puppet.

Reply:
Under what conditions will the logback configuration be re-read? That file can't be read every time that a log message is generated, can it? As it stands if you reload the JRuby pool then the [Puppetserver master class](https://github.com/puppetlabs/puppetserver/blob/master/src/ruby/puppetserver-lib/puppet/server/master.rb#L28) should be regenerated and reloaded.

Reply:
Does it need to automatically pick up the setting change, or could we document a manual step for users to perform after changing the setting?  Catching a signal (either SIGHUP, or SIGUSR1) is a common pattern I've seen for providing a way to tell a running process to re-read its configuration.  That said, signal handling is super tricky and if Puppet Server doesn't already catch those signals properly I wouldn't suggest adding code to do so at this point in the release cycle.

Reply:
Ah sure. I was under the impression you could dynamically update the log level in puppet server without having to do a server restart. In puppet you can just update the log level at any time and voila, new log level.

Reply:
That's a fair point that Puppet can internally reconfigure logging but it's a dastardly thing to do; I don't think that we need to handle cases where some extension causes Puppet to change its log level at runtime.

I did some checking and it looks like logging is only configured when a [tk app is built](https://github.com/puppetlabs/trapperkeeper/blob/master/src/puppetlabs/trapperkeeper/core.clj#L44), so I'm not entirely sure that the trapperkeeper reload logic will re-read the logback.xml. However the `initialize_puppet_server` method will be invoked each time a [JRuby instance is created](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj#L139), so I think that the Ruby logic will be called more frequently than the logback.xml file will be read.

Reply:
TK will re-initialize logback during service reload. Logback has a "scan" feature that should check the file every (configurable) 60s and update its configuration if it's changed. We've had issues with logback's scan feature working reliably (see [TK-426](https://tickets.puppetlabs.com/browse/TK-426)).

Reply:
![welp](http://i.giphy.com/yG6Wog8FTqcX6.gif)

An alternative approach is to move the log level checking into the logger itself in `#handle` if it's essential to pick up these changes, but if TK will reinitialize logback during service reload then this implementation will pick up the log level changes at that point.

Reply:
(I'm going to do some verification that reloading picks up the log level as expected.)

Reply:
You're right, finch. I was wrote that reply without seeing yours. I'm not seeing where tk re-initializes logging after a reload. (I guess I thought that [TK-424](https://tickets.puppetlabs.com/browse/TK-424) had been done).

Reply:
![](http://www.reactiongifs.com/r/2012/11/homer-pizza.gif)